game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/SelectNormalCasualties.java

```diff
@@ -0,0 +1,169 @@
+
+    } else if (hitsLeftForRestrictedTransports >= targetUnits.restrictedTransports.size()) {
+      // in addition to the combat units, all of the restricted transports were hit
+      // so kill them all without asking the player
+      targetUnits.combatUnits.addAll(targetUnits.restrictedTransports);
+      casualtyDetails = new CasualtyDetails(targetUnits.combatUnits, List.of(), true);
+
+    } else {
+      // not all restricted transports were hit so the player needs to pick which ones are killed
+      final CasualtyDetails message =
+          selectFunction.apply(
+              bridge,
+              step,
+              limitTransportsToSelect(
+                  targetUnits.restrictedTransports, hitsLeftForRestrictedTransports),
+              hitsLeftForRestrictedTransports);
+      targetUnits.combatUnits.addAll(message.getKilled());
+      casualtyDetails = new CasualtyDetails(targetUnits.combatUnits, List.of(), true);
+    }
+    return casualtyDetails;
+  }
+
+  @Value(staticConstructor = "of")
+  private static class TargetUnits {
+    List<Unit> combatUnits;
+    List<Unit> restrictedTransports;
+  }
+
+  private TargetUnits getTargetUnits(final SelectCasualties step) {
+    final TargetUnits targetUnits;
+    if (Properties.getTransportCasualtiesRestricted(step.getBattleState().getGameData())) {
+      targetUnits =
+          TargetUnits.of(
+              CollectionUtils.getMatches(
+                  step.getFiringGroup().getTargetUnits(),
+                  Matches.unitIsNotTransportButCouldBeCombatTransport().or(Matches.unitIsNotSea())),
+              CollectionUtils.getMatches(
+                  step.getFiringGroup().getTargetUnits(),
+                  Matches.unitIsTransportButNotCombatTransport().and(Matches.unitIsSea())));
+    } else {
+      targetUnits =
+          TargetUnits.of(new ArrayList<>(step.getFiringGroup().getTargetUnits()), List.of());
+    }
+    return targetUnits;
+  }
+
+  /**
+   * The maximum number of hits that this collection of units can sustain, taking into account units
+   * with two hits, and accounting for existing damage.
+   */
+  private static int getMaxHits(final Collection<Unit> units) {
+    int count = 0;
+    for (final Unit unit : units) {
+      count += UnitAttachment.get(unit.getType()).getHitPoints();
+      count -= unit.getHits();
+    }
+    return count;
+  }
+
+  /**
+   * Limit the number of transports to hitsLeftForTransports per ally
+   *
+   * @param restrictedTransports All the transports available to be selected
+   * @param hitsLeftForTransports The number of transports per ally to return
+   * @return List of transports that is equal to a maximum of hitsLeftForTransports * ally count
+   */
+  private List<Unit> limitTransportsToSelect(
+      final Collection<Unit> restrictedTransports, final int hitsLeftForTransports) {
+    final Map<GamePlayer, Collection<Unit>> alliedHitPlayer = new HashMap<>();
```

Comment:
Have you considered using Guava multimap here? That is essentially exactly the data structure of a `"map<A, Collection<B>>"`. It should simplify as you won't have to do the compute if absent, you could just add units.

EG:
```
restrictedTransports.forEach(unit -> alliedhitPlayer.add(unit.getOwner(), unit));
```